### PR TITLE
Justify write

### PR DESCRIPTION
**Pull request by strom-und-spiele**
Tends towards #36, removing some side-effects.
might have fixed some bugs that where not triggered until now.
Note that some side effects could be useful for the code working as intended, however
> Explicit is better than implicit.

### CODE REVIEW

File: src/pwm.rs [hunk 1]
```diff
@@ -446,10 +446,15 @@ macro_rules! pwm_timer_private {
...
-            tim.cr1.write(|w| w.arpe().set_bit());
+            tim.cr1.modify(|_, w| w.arpe().set_bit());
 
             // Set the "resolution" of the duty cycle (ticks before restarting at 0)
             // Oddly this is unsafe for some timers and not others
+            //
+            // FIXME: I guess this is due to some Timers not allowing values >16bit.
+            // So this needs to be checked according to the Timer type.
+            //
+            // NOTE(write): uses all bits of this register.
```
Comment (Sh3Rm4n) [inline src/pwm.rs]: Actually i think this unsafe is there because not all timers are documented in
`stm32f3`. And when they are not,this registers are marked as unsafe, because no
one programmatically confirmed that writing to this register has no side effects.

**Comment by Sh3Rm4n** [comment]
I'm personally unsure about this change. `modify` is more natural to
use than write, however this introduces additional "read from register and
store it's values" instructions. I don't know if there are optimized away when
using release, but I have not enough experience to judge this.

When initializing a peripheral (for example writing to the control register), imo using `write` is ok, because these registers weren't touched before.

Maybe, when resetting the controller, the old values in the peripheral registers are kept, and we need to cleanly write new values to it, hence the `write` invocation. But this is just a wild guess.

**Comment by strom-und-spiele** [comment]
I do agree with your objections. In some scenarios, `write` is the better choice. 

However this PR does not try to abolish `write` usage. The reason for starting this was that using `write` let to bugs in the past. The goal of this PR is to make the code more explicit.  

To give some examples:
* When writing to a stateless register, only `write` makes sense. (See, e.g. here: https://github.com/stm32-rs/stm32f3xx-hal/pull/38/files#diff-73928a829df74085c37fe73a4f3484fcR393 )
* When initializing a peripheral with a specific control register, `write` makes sure no _old_ values linger around. To make it clear to the developer what we are doing, we use `write` and add a comment.
(maybe here: https://github.com/stm32-rs/stm32f3xx-hal/pull/38/files#diff-ab68adbc36799b6f9885fbc8c64852c8R200 ? )
* When using only part of a control register, we might end up accidently disabling all other functionality (or do we want it here: https://github.com/stm32-rs/stm32f3xx-hal/pull/38/files#diff-dde82bd7fc015253d6e2e7f289dec7baR167 ? )

So to sum it up: Let's use `write` whenever it makes sense but let people know that we thought about why it makes sense.

File: src/pwm.rs (hunk 1 above)
Comment (Sh3Rm4n) [reply src/pwm.rs]: Unsafe is here, because of the reason I mentioned above. See for example [TIM17](https://stm32.agg.io/rs/STM32F303.html#TIM17:ARR) which ARR registers are still not documented in stm32f3. 

However, justifying write does still make sense. Remove the FIXME and this PR is ready to be merged :)
